components/pluralrules/src/rules/lexer.rs

```diff
@@ -0,0 +1,184 @@
+use super::ast;
+
+#[derive(Debug, PartialEq)]
+pub enum Token {
+    Operand(ast::Operand),
+    Operator(ast::Operator),
+    Number(u32),
+    Dot,
+    DotDot,
+    DotDotDot,
+    Comma,
+    Or,
+    And,
+    Modulo,
+    Junk,
+    At,
+    Decimal,
+    Integer,
+    Ellipsis,
+    Tilde,
+}
+
+#[derive(Debug)]
+pub enum LexerError {
+    ExpectedByte(u8),
+    UnknownToken(u8),
+}
+
+/// Unicode Plural Rule lexer is an iterator
+/// over tokens produced from an input string.
+///
+/// # Examples
+///
+/// ```
+/// use icu_pluralrules::rules::Lexer;
+///
+/// let input = b"i = 5";
+/// let lexer = Lexer::new(input);
+/// assert_eq!(lexer.count(), 3);
+/// ```
+pub struct Lexer<'l> {
+    chars: &'l [u8],
+    ptr: usize,
+}
+
+impl<'l> Lexer<'l> {
+    /// Constructs a new `Lexer` for a given input.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_pluralrules::rules::Lexer;
+    ///
+    /// Lexer::new(b"n = 1");
+    /// ```
+    pub fn new(input: &'l [u8]) -> Self {
+        Self {
+            chars: input,
+            ptr: 0,
+        }
+    }
+
+    fn bump(&mut self) -> Option<&u8> {
+        let ret = self.chars.get(self.ptr);
+        self.ptr += 1;
+        ret
+    }
+
+    fn take_if(&mut self, c: u8) -> bool {
+        if self.chars.get(self.ptr) == Some(&&c) {
+            self.ptr += 1;
+            true
+        } else {
+            false
+        }
+    }
+
+    fn expect(&mut self, expected: u8) -> Result<(), LexerError> {
+        if self.bump() == Some(&expected) {
+            Ok(())
+        } else {
+            Err(LexerError::ExpectedByte(expected))
+        }
+    }
+
+    fn advance_token(&mut self) -> Result<Option<Token>, LexerError> {
```

Comment:
I did not review the logic in this function in great detail.  Please let me know if you would like me to do that to double-check your algorithm correctness.

Reply:
I'm comfortable with the lexer as is. Once we have serializer I may verify round-trip behavior on all rules.